src/python/pygrackle/utilities/grdata.py

```diff
@@ -0,0 +1,774 @@
+
+    def _update(size):
+        nonlocal bar
+        if size is None and bar is not None:
+            print(flush=True)
+            bar = None
+        elif bar is not None:
+            nfill = barlen * int(size / total_bytes)
+            val = fmt.format(bar=bar, barlen=barlen, nfill=nfill, size=size / factor)
+            print(val, end="", flush=True)
+
+    return _update
+
+
+def _retrieve_url(url, dst, *, chunksize=_CHUNKSIZE):
+    """download the file from url to dst"""
+    # Online discussion about calling `response.read(chunksize)`, where `response` is a
+    # context manager object from `url.request.urlopen`, implies that there is an upper
+    # limit on the amount of data read from the http request into memory at a given
+    # point in time. This is unlikely to ever be a problem (the biggest file we
+    # currently fetch is under 10 Megabytes). However, if it does become a problem, we
+    # could just wrap the ``curl``/``wget`` cli tools.
+    try:
+        req = urllib.request.Request(url)
+        with contextlib.ExitStack() as stack:
+            out_file = stack.enter_context(open(dst, "wb"))
+            response = stack.enter_context(urllib.request.urlopen(req))
+            total_bytes = int(response.headers.get("Content-Length", -1))
+            update_progress = _progress_bar(total_bytes)
+            stack.callback(update_progress, size=None)
+
+            # write downloaded data to a file
+            downloaded_bytes = 0
+            while True:
+                update_progress(downloaded_bytes)
+                block = response.read(chunksize)
+                if not block:
+                    break
+                downloaded_bytes += len(block)
+                out_file.write(block)
+    except HTTPError as e:
+        raise GrdataError(f"server can't fulfill request to fetch {url}: {e.code}")
+    except URLError as e:
+        raise GrdataError(f"server can't be reached to fetch {url}: {e.code}")
+
+
+def _get_data_dir(system_str=None):
+    """Returns a string specifying the data directory"""
+    system_str = sys.platform if system_str is None else system_str
+    if system_str.startswith("win32"):
+        raise RuntimeError()
+    is_darwin = system_str.startswith("darwin")
+
+    # the following list specifies the order of directories in order of precedence
+    # -> defaults inspired by the API description of the platformdirs python package
+    #    from the online documentation (https://platformdirs.readthedocs.io/en/latest/).
+    #    We have NOT looked at source code!
+    # -> https://specifications.freedesktop.org/basedir-spec/latest/ instructs us to
+    #    simply skip over XDG_DATA_HOME if it specifies a relative path
+    data_dir_chain = [  # tuple fmt: (envvar, suffix_path, fatal_if_rel, is_enabled)
+        ("GRACKLE_DATA_DIR", None, True, True),
+        ("XDG_DATA_HOME", "grackle", False, True),
+        ("HOME", "Library/Application Support/grackle", True, is_darwin),
```

Comment:
Simplification opportunity: we could cut this line.

Cutting this lines means that we would treat the default directory on macOS in the exact same manner as for any other unix-like OS.